// esbuild-loader/dist/loader.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const fs_1 = __importDefault(require("fs"));
const path_1 = __importDefault(require("path"));
const esbuild_1 = require("esbuild");
const loader_utils_1 = require("loader-utils");
const joycon_1 = __importDefault(require("joycon"));
const json5_1 = __importDefault(require("json5"));
const joycon = new joycon_1.default();
joycon.addLoader({
    test: /\.json$/,
    async load(filePath) {
        try {
            const config = fs_1.default.readFileSync(filePath, 'utf8');
            return json5_1.default.parse(config);
        }
        catch (error) { // eslint-disable-line @typescript-eslint/no-implicit-any-catch
            throw new Error(`Failed to parse tsconfig at ${path_1.default.relative(process.cwd(), filePath)}: ${error.message}`);
        }
    },
});
const isTsExtensionPtrn = /\.ts$/i;
let tsConfig;
async function ESBuildLoader(source) {
    var _a, _b;
    const done = this.async();
    const options = loader_utils_1.getOptions(this);
    const transformOptions = {
        ...options,
        target: (_a = options.target) !== null && _a !== void 0 ? _a : 'es2015',
        loader: (_b = options.loader) !== null && _b !== void 0 ? _b : 'js',
        sourcemap: this.sourceMap,
        sourcefile: this.resourcePath,
    };
    if (!('tsconfigRaw' in transformOptions)) {
        if (!tsConfig) {
            tsConfig = await joycon.load(['tsconfig.json']);
        }
        if (tsConfig.data) {
            transformOptions.tsconfigRaw = tsConfig.data;
        }
    }
    // https://github.com/privatenumber/esbuild-loader/pull/107
    if (transformOptions.loader === 'tsx' &&
        isTsExtensionPtrn.test(this.resourcePath)) {
        transformOptions.loader = 'ts';
    }
    try {
        const { code, map } = await esbuild_1.transform(source, transformOptions);
        done(null, code, map && JSON.parse(map));
    }
    catch (error) {
        done(error);
    }
}
exports.default = ESBuildLoader;
