// @redwoodjs/testing/dist/MockProviders.js
"use strict";

var _Object$defineProperty = require("@babel/runtime-corejs3/core-js/object/define-property");

var _interopRequireDefault = require("@babel/runtime-corejs3/helpers/interopRequireDefault").default;

_Object$defineProperty(exports, "__esModule", {
  value: true
});

exports.MockProviders = exports.mockAuthClient = void 0;

var _react = _interopRequireDefault(require("react"));

var _auth = require("@redwoodjs/auth");

var _router = require("@redwoodjs/router");

var _apollo = require("@redwoodjs/web/apollo");

var _mockRequests = require("./mockRequests");

/**
 * NOTE: This module should not contain any nodejs functionality,
 * because it's also used by Storybook in the browser.
 */
// Import the user's Router from `./web/src/Router.{tsx,js}`,
// we pass the `children` from the user's Router to `./MockRouter.Router`
// so that we can populate the `routes object` in Storybook and tests.
const {
  default: UserRouterWithRoutes
} = require('~__REDWOOD__USER_ROUTES_FOR_MOCK');

const fakeUseAuth = () => ({
  loading: false,
  isAuthenticated: false,
  currentUser: null,
  userMetadata: null,
  logIn: async () => undefined,
  logOut: async () => undefined,
  signUp: async () => undefined,
  getToken: async () => null,
  getCurrentUser: async () => null,
  hasRole: () => false,
  reauthenticate: async () => undefined,
  client: null,
  type: 'custom',
  hasError: false
});

const mockAuthClient = {
  restoreAuthState: () => {},
  login: () => {},
  logout: () => {},
  signup: () => {},
  getToken: () => {
    return 'token';
  },
  getUserMetadata: () => {
    return _mockRequests.mockedUserMeta.currentUser;
  },
  client: 'Custom',
  type: 'custom'
};
exports.mockAuthClient = mockAuthClient;

const MockProviders = ({
  children
}) => {
  return /*#__PURE__*/_react.default.createElement(_auth.AuthProvider, {
    client: mockAuthClient,
    type: "custom"
  }, /*#__PURE__*/_react.default.createElement(_apollo.RedwoodApolloProvider, {
    useAuth: fakeUseAuth
  }, /*#__PURE__*/_react.default.createElement(UserRouterWithRoutes, null), /*#__PURE__*/_react.default.createElement(_router.LocationProvider, null, children)));
};

exports.MockProviders = MockProviders;
