// @redwoodjs/cli/dist/commands/generate/page/page.js
"use strict";

var _Object$defineProperty = require("@babel/runtime-corejs3/core-js/object/define-property");

var _interopRequireDefault = require("@babel/runtime-corejs3/helpers/interopRequireDefault").default;

_Object$defineProperty(exports, "__esModule", {
  value: true
});

exports.handler = exports.builder = exports.description = exports.command = exports.routes = exports.files = exports.paramVariants = void 0;

var _slice = _interopRequireDefault(require("@babel/runtime-corejs3/core-js/instance/slice"));

var _reduce = _interopRequireDefault(require("@babel/runtime-corejs3/core-js/instance/reduce"));

var _trim = _interopRequireDefault(require("@babel/runtime-corejs3/core-js/instance/trim"));

var _filter = _interopRequireDefault(require("@babel/runtime-corejs3/core-js/instance/filter"));

var _child_process = require("child_process");

var _camelcase = _interopRequireDefault(require("camelcase"));

var _listr = _interopRequireDefault(require("listr"));

var _pascalcase = _interopRequireDefault(require("pascalcase"));

var _internal = require("@redwoodjs/internal");

var _lib = require("../../../lib");

var _colors = _interopRequireDefault(require("../../../lib/colors"));

var _helpers = require("../helpers");

const COMPONENT_SUFFIX = 'Page';
const REDWOOD_WEB_PATH_NAME = 'pages';

const paramVariants = path => {
  var _path$match, _context, _context2;

  const param = path === null || path === void 0 ? void 0 : (_path$match = path.match(/(\{[\w:]+\})/)) === null || _path$match === void 0 ? void 0 : _path$match[1];
  const paramName = param === null || param === void 0 ? void 0 : (0, _slice.default)(_context = param.replace(/:[^}]+/, '')).call(_context, 1, -1);

  if (param === undefined) {
    return {
      propParam: '',
      propValueParam: '',
      argumentParam: '',
      paramName: '',
      paramValue: '',
      paramType: ''
    };
  } // set paramType param includes type (e.g. {id:Int}), else use string


  const paramType = param !== null && param !== void 0 && param.match(/:/) ? param === null || param === void 0 ? void 0 : (0, _slice.default)(_context2 = param.replace(/[^:]+/, '')).call(_context2, 1, -1) : 'string'; // "42" is just a value used for demonstrating parameter usage in the
  // generated page-, test-, and story-files.

  return {
    propParam: `{ ${paramName} }`,
    propValueParam: `${paramName}="42" `,
    argumentParam: `{ ${paramName}: '42' }`,
    paramName,
    paramValue: '42',
    paramType
  };
};

exports.paramVariants = paramVariants;

const files = ({
  name,
  tests,
  stories,
  typescript,
  ...rest
}) => {
  const pageFile = (0, _helpers.templateForComponentFile)({
    name,
    suffix: COMPONENT_SUFFIX,
    extension: typescript ? '.tsx' : '.js',
    webPathSection: REDWOOD_WEB_PATH_NAME,
    generator: 'page',
    templatePath: 'page.tsx.template',
    templateVars: rest
  });
  const testFile = (0, _helpers.templateForComponentFile)({
    name,
    suffix: COMPONENT_SUFFIX,
    extension: typescript ? '.test.tsx' : '.test.js',
    webPathSection: REDWOOD_WEB_PATH_NAME,
    generator: 'page',
    templatePath: 'test.tsx.template',
    templateVars: rest
  });
  const storiesFile = (0, _helpers.templateForComponentFile)({
    name,
    suffix: COMPONENT_SUFFIX,
    extension: typescript ? '.stories.tsx' : '.stories.js',
    webPathSection: REDWOOD_WEB_PATH_NAME,
    generator: 'page',
    templatePath: 'stories.tsx.template',
    templateVars: rest
  });
  const files = [pageFile];

  if (tests) {
    files.push(testFile);
  }

  if (stories) {
    files.push(storiesFile);
  } // Returns
  // {
  //    "path/to/fileA": "<<<template>>>",
  //    "path/to/fileB": "<<<template>>>",
  // }


  return (0, _reduce.default)(files).call(files, (acc, [outputPath, content]) => {
    const template = typescript ? content : (0, _lib.transformTSToJS)(outputPath, content);
    return {
      [outputPath]: template,
      ...acc
    };
  }, {});
};

exports.files = files;

const routes = ({
  name,
  path
}) => {
  return [`<Route path="${path}" page={${(0, _pascalcase.default)(name)}Page} name="${(0, _camelcase.default)(name)}" />`];
};

exports.routes = routes;
const positionalsObj = {
  path: {
    description: 'URL path to the page, or just {param}. Defaults to name',
    type: 'string'
  }
}; // @NOTE: Not exporting handler from function
// As pages need a special handler

const {
  command,
  description,
  builder
} = (0, _helpers.createYargsForComponentGeneration)({
  componentName: 'page',
  filesFn: files,
  positionalsObj
});
exports.builder = builder;
exports.description = description;
exports.command = command;

const handler = async ({
  name,
  path,
  force,
  tests,
  stories,
  typescript = false
}) => {
  var _context4;

  if (tests === undefined) {
    tests = (0, _internal.getConfig)().generate.tests;
  }

  if (stories === undefined) {
    stories = (0, _internal.getConfig)().generate.stories;
  }

  if (process.platform === 'win32') {
    // running `yarn rw g page home /` on Windows using GitBash
    // POSIX-to-Windows path conversion will kick in.
    // See https://github.com/git-for-windows/build-extra/blob/d715c9e/ReleaseNotes.md
    // As a workaround we try to detect when this has happened, and reverse
    // the action
    try {
      var _context3;

      // `cygpath -m /` will return something like 'C:/Program Files/Git/\n'
      const slashPath = (0, _trim.default)(_context3 = (0, _child_process.execSync)('cygpath -m /', {
        stdio: ['ignore', 'pipe', 'ignore']
      }).toString()).call(_context3); // `yarn rw g page home /` =>
      //   page === 'C:/Program Files/Git'
      // `yarn rw g page about /about` =>
      //   page === 'C:/Program Files/Git/about'
      // Sometimes there is a / after 'Git' to match, sometimes there isn't

      path = path.replace(new RegExp(`^${slashPath}?`), '/');
    } catch {// probably using PowerShell or cmd, in which case no special handling
      // is needed
    }
  }

  const tasks = new _listr.default((0, _filter.default)(_context4 = [{
    title: 'Generating page files...',
    task: async () => {
      path = (0, _helpers.pathName)(path, name);
      const f = await files({
        name,
        path,
        tests,
        stories,
        typescript,
        ...paramVariants(path)
      });
      return (0, _lib.writeFilesTask)(f, {
        overwriteExisting: force
      });
    }
  }, {
    title: 'Updating routes file...',
    task: async () => {
      (0, _lib.addRoutesToRouterTask)(routes({
        name,
        path: (0, _helpers.pathName)(path, name)
      }));
    }
  }]).call(_context4, Boolean), {
    collapse: false
  });

  try {
    await tasks.run();
  } catch (e) {
    console.error(_colors.default.error(e.message));
    process.exit((e === null || e === void 0 ? void 0 : e.exitCode) || 1);
  }
};

exports.handler = handler;
