// @redwoodjs/api-server/dist/handler.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.webServerHandler = exports.bothServerHandler = exports.apiServerHandler = exports.webCliOptions = exports.apiCliOptions = exports.commonOptions = void 0;

var _ansiColors = _interopRequireDefault(require("ansi-colors"));

var _internal = require("@redwoodjs/internal");

var _app = _interopRequireDefault(require("./app"));

var _withApiProxy = _interopRequireDefault(require("./middleware/withApiProxy"));

var _withFunctions = _interopRequireDefault(require("./middleware/withFunctions"));

var _withWebServer = _interopRequireDefault(require("./middleware/withWebServer"));

var _server = require("./server");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/*
 * This file has defines CLI handlers used by the redwood cli, for `rw serve`
 * Also used in index.ts for the api server
 */
const commonOptions = {
  port: {
    default: 8910,
    type: 'number',
    alias: 'p'
  },
  socket: {
    type: 'string'
  }
};
exports.commonOptions = commonOptions;
const apiCliOptions = {
  port: {
    default: 8911,
    type: 'number',
    alias: 'p'
  },
  socket: {
    type: 'string'
  },
  apiRootPath: {
    alias: ['rootPath', 'root-path'],
    default: '/',
    type: 'string',
    desc: 'Root path where your api functions are served',
    coerce: coerceRootPath
  }
};
exports.apiCliOptions = apiCliOptions;
const webCliOptions = {
  port: {
    default: 8910,
    type: 'number',
    alias: 'p'
  },
  socket: {
    type: 'string'
  },
  apiHost: {
    alias: 'api-host',
    type: 'string',
    desc: 'Forward requests from the apiProxyPath, defined in redwood.toml to this host'
  }
};
exports.webCliOptions = webCliOptions;

const apiServerHandler = async ({
  port,
  socket,
  apiRootPath
}) => {
  const tsApiServer = Date.now();
  process.stdout.write(_ansiColors.default.dim(_ansiColors.default.italic('Starting API Server... ')));
  const app = (0, _app.default)();
  const http = (0, _server.startServer)({
    port,
    socket,
    app
  }).on('listening', () => {
    console.log(_ansiColors.default.italic(_ansiColors.default.dim('Took ' + (Date.now() - tsApiServer) + ' ms')));
    const on = socket ? socket : _ansiColors.default.magenta(`http://localhost:${port}${apiRootPath}`);
    console.log(`Listening on ${on}`); // Import Server Functions.

    (0, _withFunctions.default)(app, apiRootPath);
  });
  process.on('exit', () => {
    http === null || http === void 0 ? void 0 : http.close();
    process.exit(0);
  });
};

exports.apiServerHandler = apiServerHandler;

const bothServerHandler = async ({
  port,
  socket
}) => {
  const apiRootPath = coerceRootPath((0, _internal.getConfig)().web.apiProxyPath);
  let app = (0, _app.default)(); // Attach middleware

  app = await (0, _withFunctions.default)(app, apiRootPath);
  app = (0, _withWebServer.default)(app);
  (0, _server.startServer)({
    port,
    socket,
    app
  }).on('listening', () => {
    if (socket) {
      console.log(`Listening on ${socket}`);
    }

    console.log(`Web server started on http://localhost:${port} `);
    console.log(`APIs Listening on http://localhost:${port}${apiRootPath}`);
  });
};

exports.bothServerHandler = bothServerHandler;

const webServerHandler = ({
  port,
  socket,
  apiHost
}) => {
  let app = (0, _app.default)(); // Attach middleware
  // We need to proxy api requests to prevent CORS issues

  if (apiHost) {
    const apiProxyPath = (0, _internal.getConfig)().web.apiProxyPath;
    app = (0, _withApiProxy.default)(app, {
      apiHost,
      apiProxyPath
    });
  }

  app = (0, _withWebServer.default)(app);
  (0, _server.startServer)({
    port,
    socket,
    app
  }).on('listening', () => {
    if (socket) {
      console.log(`Listening on ${socket}`);
    }

    console.log(`Web server started on http://localhost:${port} `);
  });
};

exports.webServerHandler = webServerHandler;

function coerceRootPath(path) {
  // Make sure that we create a root path that starts and ends with a slash (/)
  const prefix = path.charAt(0) !== '/' ? '/' : '';
  const suffix = path.charAt(path.length - 1) !== '/' ? '/' : '';
  return `${prefix}${path}${suffix}`;
}
