// @redwoodjs/api-server/dist/handler.js
"use strict";

var _Object$defineProperty = require("@babel/runtime-corejs3/core-js/object/define-property");

_Object$defineProperty(exports, "__esModule", {
  value: true
});

exports.handler = exports.cliOptions = void 0;

var _http = require("./http");

const cliOptions = {
  port: {
    default: 8911,
    type: 'number',
    alias: 'p'
  },
  socket: {
    type: 'string'
  },
  rootPath: {
    alias: 'root-path',
    default: '/',
    type: 'string',
    desc: 'Root path where your api functions are served',
    coerce: path => {
      // Make sure that we create a root path that starts and ends with a slash (/)
      const prefix = path.charAt(0) !== '/' ? '/' : '';
      const suffix = path.charAt(path.length - 1) !== '/' ? '/' : '';
      return `${prefix}${path}${suffix}`;
    }
  }
};
exports.cliOptions = cliOptions;

const handler = ({
  port,
  socket,
  rootPath
}) => {
  (0, _http.http)({
    port,
    socket,
    rootPath
  }).on('listening', () => {
    if (socket) {
      console.log(`Listening on ${socket}`);
    } else {
      console.log(`Listening on http://localhost:${port}${rootPath}`);
    }

    console.log();
  });
};

exports.handler = handler;
