// @redwoodjs/api/dist/logger/index.js
"use strict";

var _Object$defineProperty = require("@babel/runtime-corejs3/core-js/object/define-property");

var _interopRequireWildcard = require("@babel/runtime-corejs3/helpers/interopRequireWildcard").default;

var _interopRequireDefault = require("@babel/runtime-corejs3/helpers/interopRequireDefault").default;

_Object$defineProperty(exports, "__esModule", {
  value: true
});

exports.handlePrismaLogging = exports.emitLogLevels = exports.defaultLogLevels = exports.createLogger = exports.defaultLoggerOptions = exports.defaultPrettyPrintOptions = exports.logLevel = exports.redactionsList = exports.prettifier = exports.isPretty = exports.isProduction = exports.isTest = exports.isDevelopment = void 0;

var _stringify = _interopRequireDefault(require("@babel/runtime-corejs3/core-js/json/stringify"));

var _map = _interopRequireDefault(require("@babel/runtime-corejs3/core-js/instance/map"));

var _forEach = _interopRequireDefault(require("@babel/runtime-corejs3/core-js/instance/for-each"));

var _pino = _interopRequireDefault(require("pino"));

var prettyPrint = _interopRequireWildcard(require("pino-pretty"));

/**
 * Determines if log environment is development
 *
 * @type {boolean}
 *
 */
const isDevelopment = process.env.NODE_ENV === 'development';
/**
 * Determines if log environment is test
 *
 * @type {boolean}
 *
 */

exports.isDevelopment = isDevelopment;
const isTest = process.env.NODE_ENV === 'test';
/**
 * Determines if log environment is production by checking if not development
 *
 * @type {boolean}
 *
 */

exports.isTest = isTest;
const isProduction = !isDevelopment && !isTest;
/**
 * Determines if logs should be prettified.
 *
 * Typically if logging to a transport stream or in production
 * logs will not be prettified.
 *
 * In fact, the pino developers note:
 * "We recommend against using pino-pretty in production,
 * and highly recommend installing pino-pretty as a development dependency."
 * See: https://github.com/pinojs/pino-pretty#programmatic-integration
 *
 * One exception to this rule may be Netlify functions logging.
 * Its function logging output readability can benefit from pretty-printing.
 */

exports.isProduction = isProduction;
const isPretty = isDevelopment;
/**
 * Defines the necessary pretty printing dependency
 */

exports.isPretty = isPretty;
const prettifier = prettyPrint;
/*
 * List of keys to redact from log
 *
 * As an array, the redact option specifies paths that should have their values redacted from any log output.
 *
 * Each path must be a string using a syntax which corresponds to JavaScript dot and bracket notation.
 *
 * If an object is supplied, three options can be specified:
 *
 *      paths (String[]): Required. An array of paths
 *      censor (String): Optional. A value to overwrite key which are to be redacted. Default: '[Redacted]'
 *      remove (Boolean): Optional. Instead of censoring the value, remove both the key and the value. Default: false
 */

exports.prettifier = prettifier;
const redactionsList = ['access_token', 'accessToken', 'DATABASE_URL', 'email', 'event.headers.authorization', 'host', 'jwt', 'JWT', 'password', 'params', 'secret'];
/**
 * Determines if log level based on environment variables and
 * development or deployment environment defaults
 *
 * Set `LOG_LEVEL` env to the desired logging level. In order of priority, available levels are:
 *
 * - 'fatal'
 * - 'error'
 * - 'warn'
 * - 'info'
 * - 'debug'
 * - 'trace'
 *
 * The logging level is a __minimum__ level. For instance if `logger.level` is `'info'` then all `'fatal'`, `'error'`, `'warn'`,
 * and `'info'` logs will be enabled.
 *
 * You can pass `'silent'` to disable logging.
 *
 * @default 'warn' in Production
 * @default 'trace' in Development
 * @default 'silent' in Test
 *
 */

exports.redactionsList = redactionsList;

const logLevel = (() => {
  if (typeof process.env.LOG_LEVEL !== 'undefined') {
    return process.env.LOG_LEVEL;
  } else if (isProduction) {
    return 'warn';
  } else if (isTest) {
    return 'silent';
  } else {
    return 'trace';
  }
})();
/**
 * Defines default options when pretty printing.
 * These can be overridden individually without losing other defaults.
 *
 * Defaults are:
 *
 * - Colorize output when pretty printing
 * - Ignore certain event attributes like hostname and pid for cleaner log statements
 * - Prefix the log output with log level
 * - Use a shorted log message that omits server name
 * - Humanize time in GMT
 * */


exports.logLevel = logLevel;
const defaultPrettyPrintOptions = {
  colorize: true,
  ignore: 'hostname,pid',
  levelFirst: true,
  messageFormat: false,
  translateTime: true
};
/**
 * Defines an opinionated base logger configuration that defines
 * how to log and what to ignore.
 *
 * @default logger options are:
 *
 * - Colorize output when pretty printing
 * - Ignore certain event attributes like hostname and pid for cleaner log statements
 * - Prefix the log output with log level
 * - Use a shorted log message that omits server name
 * - Humanize time in GMT
 * - Set the default log level in test to silent, development to trace
 *   and warn in prod
 *   Or set via LOG_LEVEL environment variable
 * - Redact the host and other keys via a set redactionList
 *
 * Pretty Printing Defaults defined in defaultPrettyPrintOptions
 *
 * @see {@link https://github.com/pinojs/pino/blob/master/docs/api.md}
 * @see {@link https://github.com/pinojs/pino-pretty}
 */

exports.defaultPrettyPrintOptions = defaultPrettyPrintOptions;
const defaultLoggerOptions = {
  prettyPrint: isPretty && defaultPrettyPrintOptions,
  prettifier: isPretty && prettifier,
  level: logLevel,
  redact: redactionsList
};
/**
 * RedwoodLoggerOptions defines custom logger options that extend those available in LoggerOptions
 * and can define a destination like a file or other supported pin log transport stream
 *
 * @typedef {Object} RedwoodLoggerOptions
 * @extends LoggerOptions
 * @property {options} LoggerOptions - options define how to log
 * @property {string | DestinationStream} destination - destination defines where to log
 * @property {boolean} showConfig - Display logger configuration on initialization
 */

exports.defaultLoggerOptions = defaultLoggerOptions;

/**
 * Creates the logger
 *
 * @param options {RedwoodLoggerOptions} - Override the default logger configuration
 * @param destination {DestinationStream} - An optional destination stream
 * @param showConfig {Boolean} - Show the logger configuration. This is off by default.
 *
 * @example
 * // Create the logger to log just at the error level
 * createLogger({ options: { level: 'error' } })
 *
 * @example
 * // Create the logger to log to a file
 * createLogger({ destination: { 'var/logs/redwood-api.log' } })
 *
 * @return {BaseLogger} - The configured logger
 */
const createLogger = ({
  options,
  destination,
  showConfig = false
}) => {
  const hasDestination = typeof destination !== 'undefined';
  const isFile = hasDestination && typeof destination === 'string';
  const isStream = hasDestination && !isFile;
  const stream = destination; // override, but retain default pretty print options

  if (isPretty && options && options.prettyPrint) {
    const prettyOptions = {
      prettyPrint: { ...defaultLoggerOptions.prettyPrint,
        ...options.prettyPrint
      }
    };
    delete options.prettyPrint;
    options = { ...defaultLoggerOptions,
      ...prettyOptions,
      ...options
    };
  } else {
    options = { ...defaultLoggerOptions,
      ...options
    };
  }

  if (showConfig) {
    console.log('Logger Configuration');
    console.log(`isProduction: ${isProduction}`);
    console.log(`isDevelopment: ${isDevelopment}`);
    console.log(`isTest: ${isTest}`);
    console.log(`isPretty: ${isPretty}`);
    console.log(`isFile: ${isFile}`);
    console.log(`isStream: ${isStream}`);
    console.log(`logLevel: ${logLevel}`);
    console.log(`options: ${(0, _stringify.default)(options, null, 2)}`);
    console.log(`destination: ${destination}`);
  }

  if (isFile) {
    if (isProduction) {
      console.warn('Please make certain that file system access is available when logging to a file in a production environment.');
    }

    return (0, _pino.default)(options, stream);
  } else {
    if (isStream && isDevelopment && !isTest) {
      console.warn('Logs will be sent to the transport stream in the current development environment.');
    }

    if (isStream && options.prettyPrint) {
      console.warn('Logs sent to the transport stream are being prettified. This format may be incompatible.');
    }

    return (0, _pino.default)(options, stream);
  }
};
/**
 * To help you identify particularly slow parameter values,
 * For example, Heroku outputs the slowest queries (that take 2 seconds or more)
 **/


exports.createLogger = createLogger;
const SLOW_QUERY_THRESHOLD = 2_000; // 2 seconds

/**
 * Determines the type and level of logging.
 *
 * @see {@link https://www.prisma.io/docs/reference/api-reference/prisma-client-reference#log}
 */

const defaultLogLevels = ['info', 'warn', 'error'];
/**
 * Generates the Prisma Log Definitions for the Prisma Client to emit
 *
 * @return Prisma.LogDefinition[]
 */

exports.defaultLogLevels = defaultLogLevels;

const emitLogLevels = setLogLevels => {
  return setLogLevels === null || setLogLevels === void 0 ? void 0 : (0, _map.default)(setLogLevels).call(setLogLevels, level => {
    return {
      emit: 'event',
      level
    };
  });
};
/**
 * Defines what is needed to setup Prisma Client logging used in handlePrismaLogging
 *
 *
 * @param db {PrismaClient} - The Prisma Client instance
 * @param logger {BaseLogger} - The Redwood logger instance
 * @param logLevels {LogLevel[]} - The log levels . Should match those set with emitLogLevels
 *
 * @see emitLogLevels
 *
 */


exports.emitLogLevels = emitLogLevels;

/**
 * Sets up event-based logging on the Prisma client
 *
 * Sets emit to event for a specific log level, such as query
 * using the $on() method to subscribe to the event
 *
 * @see {@link https://www.prisma.io/docs/concepts/components/prisma-client/working-with-prismaclient/logging}
 *
 * @param PrismaLoggingConfig
 *
 * @example
 *
 * handlePrismaLogging({
 *  db,
 *  logger,
 *  logLevels: ['info', 'warn', 'error'],
 * })
 *
 * @return {void}
 *
 */
const handlePrismaLogging = config => {
  var _config$logLevels;

  const logger = config.logger.child({
    prisma: {
      clientVersion: config.db['_clientVersion']
    }
  });
  (_config$logLevels = config.logLevels) === null || _config$logLevels === void 0 ? void 0 : (0, _forEach.default)(_config$logLevels).call(_config$logLevels, level => {
    if (level === 'query') {
      config.db.$on(level, queryEvent => {
        if (queryEvent.duration >= SLOW_QUERY_THRESHOLD) {
          logger.warn({ ...queryEvent
          }, `Slow Query performed in ${queryEvent.duration} msec`);
        } else {
          logger.debug({ ...queryEvent
          }, `Query performed in ${queryEvent.duration} msec`);
        }
      });
    } else {
      config.db.$on(level, logEvent => {
        switch (level) {
          case 'info':
            logger.info({ ...logEvent
            }, logEvent.message);
            break;

          case 'warn':
            logger.warn({ ...logEvent
            }, logEvent.message);
            break;

          case 'error':
            logger.error({ ...logEvent
            }, logEvent.message);
            break;

          default:
            logger.info({ ...logEvent
            }, logEvent.message);
        }
      });
    }
  });
  return;
};

exports.handlePrismaLogging = handlePrismaLogging;
