// @redwoodjs/structure/dist/model/RWSDLField.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.RWSDLField = void 0;

var _vscodeLanguageserverTypes = require("vscode-languageserver-types");

var _errors = require("../errors");

var _ide = require("../ide");

var _decorators = require("../x/decorators");

var _URL = require("../x/URL");

var _vscodeLanguageserverTypes2 = require("../x/vscode-languageserver-types");

var _dec, _dec2, _dec3, _dec4, _dec5, _dec6, _dec7, _class;

function _applyDecoratedDescriptor(target, property, decorators, descriptor, context) { var desc = {}; Object.keys(descriptor).forEach(function (key) { desc[key] = descriptor[key]; }); desc.enumerable = !!desc.enumerable; desc.configurable = !!desc.configurable; if ('value' in desc || desc.initializer) { desc.writable = true; } desc = decorators.slice().reverse().reduce(function (desc, decorator) { return decorator(target, property, desc) || desc; }, desc); if (context && desc.initializer !== void 0) { desc.value = desc.initializer ? desc.initializer.call(context) : void 0; desc.initializer = undefined; } if (desc.initializer === void 0) { Object.defineProperty(target, property, desc); desc = null; } return desc; }

let RWSDLField = (_dec = (0, _decorators.lazy)(), _dec2 = (0, _decorators.lazy)(), _dec3 = (0, _decorators.lazy)(), _dec4 = (0, _decorators.lazy)(), _dec5 = (0, _decorators.lazy)(), _dec6 = (0, _decorators.lazy)(), _dec7 = (0, _decorators.lazy)(), (_class = class RWSDLField extends _ide.BaseNode {
  constructor(objectTypeDef, field, parent) {
    super();
    this.objectTypeDef = objectTypeDef;
    this.field = field;
    this.parent = parent;
  }

  get id() {
    return this.parent.id + ' ' + this.objectTypeDef.name.value + '.' + this.name;
  }
  /**
   * The location of this field.
   * Calculating this is slightly complicated since it is embedded within a TaggedTemplateLiteral
   */


  get location() {
    let {
      start,
      end
    } = this.field.loc;
    const node = this.parent.schemaStringNode;
    start += node.getPos() + 1; // we add one to account for the quote (`)

    end += node.getPos() + 1;
    const startPos = (0, _vscodeLanguageserverTypes2.Position_fromTSMorphOffset)(start, node.getSourceFile());
    const endPos = (0, _vscodeLanguageserverTypes2.Position_fromTSMorphOffset)(end, node.getSourceFile());
    return {
      uri: this.parent.uri,
      range: {
        start: startPos,
        end: endPos
      }
    };
  }

  get name() {
    return this.field.name.value;
  }

  get argumentNames() {
    var _this$field$arguments;

    return ((_this$field$arguments = this.field.arguments) !== null && _this$field$arguments !== void 0 ? _this$field$arguments : []).map(a => a.name.value);
  }

  *ideInfo() {
    if (this.impl) {
      yield {
        kind: 'Implementation',
        location: this.location,
        target: (0, _vscodeLanguageserverTypes2.Location_fromNode)(this.impl.node)
      };
    }
  }
  /**
   * TODO: describe in prose what is going on here.
   * this is an important rule
   */


  get impl() {
    var _this$parent$service$, _this$parent$service;

    return ((_this$parent$service$ = (_this$parent$service = this.parent.service) === null || _this$parent$service === void 0 ? void 0 : _this$parent$service.funcs) !== null && _this$parent$service$ !== void 0 ? _this$parent$service$ : []).find(f => f.name === this.name);
  } // TODO: improve snippet


  get defaultImplSnippet() {
    var _this$field$arguments2;

    const args = (_this$field$arguments2 = this.field.arguments) !== null && _this$field$arguments2 !== void 0 ? _this$field$arguments2 : [];
    const params = args.map(a => a.name.value).join(',');
    return `
export const ${this.field.name.value} = ({${params}}) => {
  // TODO: implement
}`;
  }

  get quickFix_addImplementation() {
    const {
      service
    } = this.parent;
    const change = new _vscodeLanguageserverTypes.WorkspaceChange({
      documentChanges: []
    });

    let insertPosition = _vscodeLanguageserverTypes.Position.create(0, 0);

    let uri = (0, _URL.URL_file)(this.parent.serviceFilePath);

    if (service) {
      // we'll insert into the end of an existing file
      const lastLine = service.sf.getEndLineNumber();
      insertPosition = _vscodeLanguageserverTypes.Position.create(lastLine, 0);
      uri = service.uri;
    } else {
      // file doesn't exist
      // create the service file before inserting
      change.createFile(uri);
    } // insert


    change.getTextEditChange({
      uri,
      version: null
    }).insert(insertPosition, this.defaultImplSnippet);
    return {
      title: 'Add implementation',
      edit: change.edit
    };
  }

  *diagnostics() {
    if (!this.impl) {
      const {
        uri,
        range
      } = this.location;
      yield {
        uri,
        diagnostic: {
          range,
          message: 'Service Not Implemented',
          severity: _vscodeLanguageserverTypes.DiagnosticSeverity.Error,
          code: _errors.RWError.SERVICE_NOT_IMPLEMENTED
        },
        quickFix: async () => this.quickFix_addImplementation
      };
    }
  }

}, (_applyDecoratedDescriptor(_class.prototype, "id", [_dec], Object.getOwnPropertyDescriptor(_class.prototype, "id"), _class.prototype), _applyDecoratedDescriptor(_class.prototype, "location", [_dec2], Object.getOwnPropertyDescriptor(_class.prototype, "location"), _class.prototype), _applyDecoratedDescriptor(_class.prototype, "name", [_dec3], Object.getOwnPropertyDescriptor(_class.prototype, "name"), _class.prototype), _applyDecoratedDescriptor(_class.prototype, "argumentNames", [_dec4], Object.getOwnPropertyDescriptor(_class.prototype, "argumentNames"), _class.prototype), _applyDecoratedDescriptor(_class.prototype, "impl", [_dec5], Object.getOwnPropertyDescriptor(_class.prototype, "impl"), _class.prototype), _applyDecoratedDescriptor(_class.prototype, "defaultImplSnippet", [_dec6], Object.getOwnPropertyDescriptor(_class.prototype, "defaultImplSnippet"), _class.prototype), _applyDecoratedDescriptor(_class.prototype, "quickFix_addImplementation", [_dec7], Object.getOwnPropertyDescriptor(_class.prototype, "quickFix_addImplementation"), _class.prototype)), _class));
exports.RWSDLField = RWSDLField;
