// tailwindcss-hero-patterns/.history/tailwind.config_20200528151008.js
const plugin = require("tailwindcss/plugin");
const patterns = require("./patterns");

module.exports = {
  theme: {
    extend: {
      colors: {
        // palette generated with https://javisperez.github.io/tailwindcolorshades/#/?Tolopea=2b0447&tv=1
        primary: {
          default: "#2b0447",
          "100": "#f5f5f5",
          "200": "#eeeeee",
          "300": "#e0e0e0",
          "400": "#bdbdbd",
          "500": "#9e9e9e",
          "600": "#757575",
          "700": "#616161",
          "800": "#424242",
          "900": "#2b0447",
        },
      },
    },
  },
  variants: {},

  plugins: [
    require("tailwindcss-gradients"),
    plugin(function ({ addUtilities, theme }) {
      const colors = theme("colors", {});

      // flatten colors
      const flattenColors = {};
      Object.entries(colors).map(([name, color]) => {
        if (typeof color === "object") {
          Object.entries(color).map(([nestedName, nestedColor]) => {
            flattenColors[
              `${name}${nestedName === "default" ? "" : `-${nestedName}`}`
            ] = nestedColor;
          });
        } else {
          flattenColors[name] = color;
        }
      });

      let newUtilities = {};
      Object.entries(patterns).map(([name, pattern]) =>
        Object.entries(flattenColors).map(
          ([colorName, color]) =>
            (newUtilities[`.svg-${name}-${colorName}`] = {
              backgroundImage: pattern
                .replace("{{color}}", color.replace("#", "%23"))
                .replace("{{opacity}}", 1),
            })
        )
      );

      addUtilities(newUtilities);
    }),
  ],
};
