// tailwindcss-hero-patterns/.history/tailwind.config_20200529063129.js
module.exports = {
  purge: ["./docs/index.html"],
  theme: {
    extend: {
      colors: {
        // palette generated with https://javisperez.github.io/tailwindcolorshades/#/?Tolopea=2b0447&tv=1
        primary: {
          default: "#2b0447",
          "100": "#f5f5f5",
          "200": "#eeeeee",
          "300": "#e0e0e0",
          "400": "#bdbdbd",
          "500": "#9e9e9e",
          "600": "#757575",
          "700": "#616161",
          "800": "#424242",
          "900": "#2b0447",
        },
      },
      linearGradientDirections: {
        r: "to right",
      },
      linearGradientColors: {
        "transparent-to-primary": ["#bdbdbd 0%", "#bdbdbd 72%", "#bdbdbd 100%"],
        "transparent-to-red": [
          "rgba(0,0,0,0) 0%",
          "red-400 72%",
          "red-400 100%",
        ],
        "transparent-to-blue": [
          "rgba(0,0,0,0) 0%",
          "rgba(99, 179, 237,1) 72%",
          "rgba(99, 179, 237,1) 100%",
        ],
      },
      heropatterns: {
        ellipsis: `<svg height="140" width="500"><ellipse cx="200" cy="80" rx="100" ry="50" style="fill:{{color}};stroke:purple;stroke-width:2" /></svg>`,
      },
    },
  },
  variants: {},
  plugins: [require("tailwindcss-gradients"), require("./src")],
};
