// tailwindcss-hero-patterns/.history/tailwind.config_20200529105048.js
const patterns = require("./src/patterns");

function getGradient(theme, color) {
  const hex = theme("colors")[color][400];
  return ["transparent 0%", `${hex} 72%`, `${hex} 100%`];
}

const colors = [
  "gray",
  "red",
  "orange",
  "yellow",
  "green",
  "teal",
  "blue",
  "indigo",
  "purple",
  "pink",
  "primary",
];

module.exports = {
  purge: ["./docs/index.html"],
  theme: {
    heropatterns: {
      architect: patterns.architect,
    },
    extend: {
      colors: {
        // palette generated with https://javisperez.github.io/tailwindcolorshades/#/?Tolopea=2b0447&tv=1
        primary: {
          default: "#2b0447",
          "100": "#f5f5f5",
          "200": "#eeeeee",
          "300": "#e0e0e0",
          "400": "#bdbdbd",
          "500": "#9e9e9e",
          "600": "#757575",
          "700": "#616161",
          "800": "#424242",
          "900": "#2b0447",
        },
      },
      linearGradientDirections: {
        r: "to right",
      },
      linearGradientColors: (theme) =>
        colors.reduce((accum, color) => {
          accum[`transparent-to-${color}`] = getGradient(theme, color);
          return accum;
        }, {}),
      heropatterns: {
        ellipsis: `url("data:image/svg+xml;charset=utf8,%3C?xml version='1.0' encoding='utf-8'?%3E%3C!DOCTYPE svg PUBLIC '-//W3C//DTD SVG 1.1//EN' 'http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd'%3E%3Csvg version='1.1' id='Layer_1' xmlns='http://www.w3.org/2000/svg' xmlns:xlink='http://www.w3.org/1999/xlink' x='0px' y='0px' viewBox='0 0 2560 1600' enable-background='new 0 0 2560 1600' xml:space='preserve'%3E%3Cpattern width='300' height='300' patternUnits='userSpaceOnUse' id='Polka_Dot_Pattern' viewBox='0 -300 300 300' overflow='visible'%3E%3Cg%3E%3Cpolygon fill='none' points='0,-300 300,-300 300,0 0,0 '/%3E%3Cpath fill='%238367FF' d='M100-225c0,13.8071-11.1929,25-25,25s-25-11.1929-25-25s11.1929-25,25-25S100-238.8071,100-225z'/%3E%3Cpath fill='%23FF0000' d='M100-75c0,13.8071-11.1929,25-25,25S50-61.1929,50-75s11.1929-25,25-25S100-88.8071,100-75z'/%3E%3Cpath fill='%23F1FF00' d='M250-225c0,13.8071-11.1929,25-25,25s-25-11.1929-25-25s11.1929-25,25-25S250-238.8071,250-225z'/%3E%3Cpath fill='%23FF00FF' d='M250-75c0,13.8071-11.1929,25-25,25s-25-11.1929-25-25s11.1929-25,25-25S250-88.8071,250-75z'/%3E%3C/g%3E%3C/pattern%3E%3Crect fill='url(%23Polka_Dot_Pattern)' width='2560' height='1600'/%3E%3C/svg%3E");`,
      },
    },
  },
  variants: {},
  plugins: [require("tailwindcss-gradients"), require("./src")],
};
