// tailwindcss-hero-patterns/.history/tailwind.config_20200528144503.js
const plugin = require("tailwindcss/plugin");
const patterns = require("./patterns");

module.exports = {
  theme: {
    extend: {
      colors: {
        // palette generated with https://javisperez.github.io/tailwindcolorshades/#/?Tolopea=2b0447&tv=1
        primary: {
          default: "#2b0447",
          "100": "#f5f5f5",
          "200": "#eeeeee",
          "300": "#e0e0e0",
          "400": "#bdbdbd",
          "500": "#9e9e9e",
          "600": "#757575",
          "700": "#616161",
          "800": "#424242",
          "900": "#2b0447",
        },
      },
    },

    linearGradientDirections: {
      // defaults to these values
      t: "to top",
      tr: "to top right",
      r: "to right",
      br: "to bottom right",
      b: "to bottom",
      bl: "to bottom left",
      l: "to left",
      tl: "to top left",
    },
    linearGradientColors: {
      // defaults to {}
      red: "#f00",
      "transparent-blue": ["transparent", "#00f"],
      "red-blue": ["#f00", "#00f"],
      "red-green-blue": ["#f00", "#0f0", "#00f"],
      "black-white-with-stops": ["#000", "#000 45%", "#fff 55%", "#fff"],
    },
    radialGradientShapes: {
      // defaults to this value
      default: "ellipse",
    },
    radialGradientSizes: {
      // defaults to this value
      default: "closest-side",
    },
    radialGradientPositions: {
      // defaults to these values
      default: "center",
      t: "top",
      tr: "top right",
      r: "right",
      br: "bottom right",
      b: "bottom",
      bl: "bottom left",
      l: "left",
      tl: "top left",
    },
    radialGradientColors: {
      // defaults to {}
      red: "#f00",
      "red-blue": ["#f00", "#00f"],
      "red-green-blue": ["#f00", "#0f0", "#00f"],
      "black-white-with-stops": ["#000", "#000 45%", "#fff 55%", "#fff"],
    },
    conicGradientStartingAngles: {
      // defaults to this value
      default: "0",
    },
    conicGradientPositions: {
      // defaults to these values
      default: "center",
      t: "top",
      tr: "top right",
      r: "right",
      br: "bottom right",
      b: "bottom",
      bl: "bottom left",
      l: "left",
      tl: "top left",
    },
    conicGradientColors: {
      // defaults to {}
      red: "#f00",
      "red-blue": ["#f00", "#00f"],
      "red-green-blue": ["#f00", "#0f0", "#00f"],
      checkerboard: [
        "white 90deg",
        "black 90deg 180deg",
        "white 180deg 270deg",
        "black 270deg",
      ],
    },
    repeatingLinearGradientDirections: (theme) =>
      theme("linearGradientDirections"), // defaults to this value
    repeatingLinearGradientColors: (theme) => theme("linearGradientColors"), // defaults to {}
    repeatingLinearGradientLengths: {
      // defaults to {}
      sm: "25px",
      md: "50px",
      lg: "100px",
    },
    repeatingRadialGradientShapes: (theme) => theme("radialGradientShapes"), // defaults to this value
    repeatingRadialGradientSizes: {
      // defaults to this value
      default: "farthest-corner",
    },
    repeatingRadialGradientPositions: (theme) =>
      theme("radialGradientPositions"), // defaults to this value
    repeatingRadialGradientColors: (theme) => theme("radialGradientColors"), // defaults to {}
    repeatingRadialGradientLengths: {
      // defaults to {}
      sm: "25px",
      md: "50px",
      lg: "100px",
    },
    repeatingConicGradientStartingAngles: (theme) =>
      theme("conicGradientStartingAngles"), // defaults to this value
    repeatingConicGradientPositions: (theme) => theme("conicGradientPositions"), // defaults to this value
    repeatingConicGradientColors: {
      // defaults to {}
      red: "#f00",
      "red-blue": ["#f00", "#00f"],
      "red-green-blue": ["#f00", "#0f0", "#00f"],
      starburst: ["white 0 5deg", "blue 5deg"],
    },
    repeatingConicGradientLengths: {
      // defaults to {}
      sm: "10deg",
      md: "20deg",
      lg: "40deg",
    },
  },
  variants: {
    // all the following default to ['responsive']
    backgroundImage: ["responsive"], // this is for the "bg-none" utility
    linearGradients: ["responsive"],
    radialGradients: ["responsive"],
    conicGradients: ["responsive"],
    repeatingLinearGradients: ["responsive"],
    repeatingRadialGradients: ["responsive"],
    repeatingConicGradients: ["responsive"],
  },

  plugins: [
    require("tailwindcss-gradients"),
    plugin(function ({ addUtilities, theme }) {
      const colors = theme("colors", {});

      // flatten colors
      const flattenColors = {};
      Object.entries(colors).map(([name, color]) => {
        if (typeof color === "object") {
          Object.entries(color).map(([nestedName, nestedColor]) => {
            flattenColors[
              `${name}${nestedName === "default" ? "" : `-${nestedName}`}`
            ] = nestedColor;
          });
        } else {
          flattenColors[name] = color;
        }
      });

      let newUtilities = {};
      Object.entries(patterns).map(([name, pattern]) =>
        Object.entries(flattenColors).map(
          ([colorName, color]) =>
            (newUtilities[`.svg-${name}-${colorName}`] = {
              backgroundImage: pattern
                .replace("{{color}}", color.replace("#", "%23"))
                .replace("{{opacity}}", 1),
            })
        )
      );

      addUtilities(newUtilities);
    }),
  ],
};
