// tailwindcss-hero-patterns/.history/tailwind.config_20200529104602.js
const patterns = require("./src/patterns");

function getGradient(theme, color) {
  const hex = theme("colors")[color][400];
  return ["transparent 0%", `${hex} 72%`, `${hex} 100%`];
}

const colors = [
  "gray",
  "red",
  "orange",
  "yellow",
  "green",
  "teal",
  "blue",
  "indigo",
  "purple",
  "pink",
  "primary",
];

module.exports = {
  purge: ["./docs/index.html"],
  theme: {
    heropatterns: {
      architect: patterns.architect,
    },
    extend: {
      colors: {
        // palette generated with https://javisperez.github.io/tailwindcolorshades/#/?Tolopea=2b0447&tv=1
        primary: {
          default: "#2b0447",
          "100": "#f5f5f5",
          "200": "#eeeeee",
          "300": "#e0e0e0",
          "400": "#bdbdbd",
          "500": "#9e9e9e",
          "600": "#757575",
          "700": "#616161",
          "800": "#424242",
          "900": "#2b0447",
        },
      },
      linearGradientDirections: {
        r: "to right",
      },
      linearGradientColors: (theme) =>
        colors.reduce((accum, color) => {
          accum[`transparent-to-${color}`] = getGradient(theme, color);
          return accum;
        }, {}),
      heropatterns: {
        ellipsis: `url("data:image/svg+xml,%3Csvg+height%3D%22140%22+width%3D%22500%22%3E%3Cellipse+cx%3D%22200%22+cy%3D%2280%22+rx%3D%22100%22+ry%3D%2250%22+style%3D%22fill%3Ayellow%3Bstroke%3Apurple%3Bstroke-width%3A2%22+%2F%3E%3C%2Fsvg%3E")`,
      },
    },
  },
  variants: {},
  plugins: [require("tailwindcss-gradients"), require("./src")],
};
