// tailwindcss-hero-patterns/.history/tailwind.config_20200528080250.js
const plugin = require("tailwindcss/plugin");

module.exports = {
  theme: {
    extend: {
      colors: {
        // palette generated with https://javisperez.github.io/tailwindcolorshades/#/?Tolopea=2b0447&tv=1
        primary: {
          default: "#2b0447",
          "100": "#f5f5f5",
          "200": "#eeeeee",
          "300": "#e0e0e0",
          "400": "#bdbdbd",
          "500": "#9e9e9e",
          "600": "#757575",
          "700": "#616161",
          "800": "#424242",
          "900": "#2b0447",
        },
      },
    },

    linearGradientDirections: {
      // defaults to these values
      t: "to top",
      tr: "to top right",
      r: "to right",
      br: "to bottom right",
      b: "to bottom",
      bl: "to bottom left",
      l: "to left",
      tl: "to top left",
    },
    linearGradientColors: {
      // defaults to {}
      red: "#f00",
      "red-blue": ["#f00", "#00f"],
      "red-green-blue": ["#f00", "#0f0", "#00f"],
      "black-white-with-stops": ["#000", "#000 45%", "#fff 55%", "#fff"],
    },
    radialGradientShapes: {
      // defaults to this value
      default: "ellipse",
    },
    radialGradientSizes: {
      // defaults to this value
      default: "closest-side",
    },
    radialGradientPositions: {
      // defaults to these values
      default: "center",
      t: "top",
      tr: "top right",
      r: "right",
      br: "bottom right",
      b: "bottom",
      bl: "bottom left",
      l: "left",
      tl: "top left",
    },
    radialGradientColors: {
      // defaults to {}
      red: "#f00",
      "red-blue": ["#f00", "#00f"],
      "red-green-blue": ["#f00", "#0f0", "#00f"],
      "black-white-with-stops": ["#000", "#000 45%", "#fff 55%", "#fff"],
    },
    conicGradientStartingAngles: {
      // defaults to this value
      default: "0",
    },
    conicGradientPositions: {
      // defaults to these values
      default: "center",
      t: "top",
      tr: "top right",
      r: "right",
      br: "bottom right",
      b: "bottom",
      bl: "bottom left",
      l: "left",
      tl: "top left",
    },
    conicGradientColors: {
      // defaults to {}
      red: "#f00",
      "red-blue": ["#f00", "#00f"],
      "red-green-blue": ["#f00", "#0f0", "#00f"],
      checkerboard: [
        "white 90deg",
        "black 90deg 180deg",
        "white 180deg 270deg",
        "black 270deg",
      ],
    },
    repeatingLinearGradientDirections: (theme) =>
      theme("linearGradientDirections"), // defaults to this value
    repeatingLinearGradientColors: (theme) => theme("linearGradientColors"), // defaults to {}
    repeatingLinearGradientLengths: {
      // defaults to {}
      sm: "25px",
      md: "50px",
      lg: "100px",
    },
    repeatingRadialGradientShapes: (theme) => theme("radialGradientShapes"), // defaults to this value
    repeatingRadialGradientSizes: {
      // defaults to this value
      default: "farthest-corner",
    },
    repeatingRadialGradientPositions: (theme) =>
      theme("radialGradientPositions"), // defaults to this value
    repeatingRadialGradientColors: (theme) => theme("radialGradientColors"), // defaults to {}
    repeatingRadialGradientLengths: {
      // defaults to {}
      sm: "25px",
      md: "50px",
      lg: "100px",
    },
    repeatingConicGradientStartingAngles: (theme) =>
      theme("conicGradientStartingAngles"), // defaults to this value
    repeatingConicGradientPositions: (theme) => theme("conicGradientPositions"), // defaults to this value
    repeatingConicGradientColors: {
      // defaults to {}
      red: "#f00",
      "red-blue": ["#f00", "#00f"],
      "red-green-blue": ["#f00", "#0f0", "#00f"],
      starburst: ["white 0 5deg", "blue 5deg"],
    },
    repeatingConicGradientLengths: {
      // defaults to {}
      sm: "10deg",
      md: "20deg",
      lg: "40deg",
    },
  },
  variants: {
    // all the following default to ['responsive']
    backgroundImage: ["responsive"], // this is for the "bg-none" utility
    linearGradients: ["responsive"],
    radialGradients: ["responsive"],
    conicGradients: ["responsive"],
    repeatingLinearGradients: ["responsive"],
    repeatingRadialGradients: ["responsive"],
    repeatingConicGradients: ["responsive"],
  },
  plugins: [
    require("tailwindcss-gradients"),

    plugin(function ({ addUtilities, addComponents, e, prefix, config }) {
      // Add your custom styles here
      const newUtilities = {
        ".sven": {
          "background-image": `url('data:image/svg+xml,%3Csvg xmlns=\'http://www.w3.org/2000/svg\' width=\'40\' height=\'40\' viewBox=\'0 0 40 40\'%3E%3Cg fill-rule=\'evenodd\'%3E%3Cg fill=\'%239C92AC\' fill-opacity=\'0.4\'%3E%3Cpath d=\'M0 38.59l2.83-2.83 1.41 1.41L1.41 40H0v-1.41zM0 1.4l2.83 2.83 1.41-1.41L1.41 0H0v1.41zM38.59 40l-2.83-2.83 1.41-1.41L40 38.59V40h-1.41zM40 1.41l-2.83 2.83-1.41-1.41L38.59 0H40v1.41zM20 18.6l2.83-2.83 1.41 1.41L21.41 20l2.83 2.83-1.41 1.41L20 21.41l-2.83 2.83-1.41-1.41L18.59 20l-2.83-2.83 1.41-1.41L20 18.59z\'/%3E%3C/g%3E%3C/g%3E%3C/svg%3E')`,
        },
        ".skew-15deg": {
          transform: "skewY(-15deg)",
        },
      };

      addUtilities(newUtilities);
    }),
  ],
};
