// tailwindcss-hero-patterns/.history/tailwind.config_20200529104348.js
const patterns = require("./src/patterns");

function getGradient(theme, color) {
  const hex = theme("colors")[color][400];
  return ["transparent 0%", `${hex} 72%`, `${hex} 100%`];
}

const colors = [
  "gray",
  "red",
  "orange",
  "yellow",
  "green",
  "teal",
  "blue",
  "indigo",
  "purple",
  "pink",
  "primary",
];

module.exports = {
  purge: ["./docs/index.html"],
  heropatterns: {
    architect: patterns.architect,
  },
  theme: {
    extend: {
      colors: {
        // palette generated with https://javisperez.github.io/tailwindcolorshades/#/?Tolopea=2b0447&tv=1
        primary: {
          default: "#2b0447",
          "100": "#f5f5f5",
          "200": "#eeeeee",
          "300": "#e0e0e0",
          "400": "#bdbdbd",
          "500": "#9e9e9e",
          "600": "#757575",
          "700": "#616161",
          "800": "#424242",
          "900": "#2b0447",
        },
      },
      linearGradientDirections: {
        r: "to right",
      },
      linearGradientColors: (theme) =>
        colors.reduce((accum, color) => {
          accum[`transparent-to-${color}`] = getGradient(theme, color);
          return accum;
        }, {}),
      heropatterns: {
        ellipsis: `url("data:image/svg+xml,<svg width='84' height='48' viewBox='0 0 84 48' xmlns='http://www.w3.org/2000/svg'><path d='M0 0h12v6H0V0zm28 8h12v6H28V8zm14-8h12v6H42V0zm14 0h12v6H56V0zm0 8h12v6H56V8zM42 8h12v6H42V8zm0 16h12v6H42v-6zm14-8h12v6H56v-6zm14 0h12v6H70v-6zm0-16h12v6H70V0zM28 32h12v6H28v-6zM14 16h12v6H14v-6zM0 24h12v6H0v-6zm0 8h12v6H0v-6zm14 0h12v6H14v-6zm14 8h12v6H28v-6zm-14 0h12v6H14v-6zm28 0h12v6H42v-6zm14-8h12v6H56v-6zm0-8h12v6H56v-6zm14 8h12v6H70v-6zm0 8h12v6H70v-6zM14 24h12v6H14v-6zm14-8h12v6H28v-6zM14 8h12v6H14V8zM0 8h12v6H0V8z' fill='#fff5f5' fill-opacity='1' fill-rule='evenodd'/></svg>")`,
      },
    },
  },
  variants: {},
  plugins: [require("tailwindcss-gradients"), require("./src")],
};
