// tailwindcss-hero-patterns/.history/tailwind.config_20200529064319.js
function getGradient(theme, color) {
  const hex = theme("colors")[color][400];
  return ["transparent 0%", `${hex} 72%`, `${hex} 100%`];
}

const colors = [
  "gray",
  "red",
  "orange",
  "yellow",
  "green",
  "teal",
  "blue",
  "indigo",
  "purple",
  "pink",
  "primary",
];

module.exports = {
  purge: ["./docs/index.html"],
  theme: {
    extend: {
      colors: {
        // palette generated with https://javisperez.github.io/tailwindcolorshades/#/?Tolopea=2b0447&tv=1
        primary: {
          default: "#2b0447",
          "100": "#f5f5f5",
          "200": "#eeeeee",
          "300": "#e0e0e0",
          "400": "#bdbdbd",
          "500": "#9e9e9e",
          "600": "#757575",
          "700": "#616161",
          "800": "#424242",
          "900": "#2b0447",
        },
      },
      linearGradientDirections: {
        r: "to right",
      },
      linearGradientColors: (theme) =>
        colors.reduce((accum, color) => {
          accum[`transparent-to-${color}`] = getGradient(theme, "red");
          return accum;
        }, {}),
      heropatterns: {
        ellipsis: `<svg height="140" width="500"><ellipse cx="200" cy="80" rx="100" ry="50" style="fill:{{color}};stroke:purple;stroke-width:2" /></svg>`,
      },
    },
  },
  variants: {},
  plugins: [require("tailwindcss-gradients"), require("./src")],
};
