// @redwoodjs/cli/dist/commands/build.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.handler = exports.builder = exports.description = exports.command = void 0;

var _fs = _interopRequireDefault(require("fs"));

var _execa = _interopRequireDefault(require("execa"));

var _listr = _interopRequireDefault(require("listr"));

var _listrVerboseRenderer = _interopRequireDefault(require("listr-verbose-renderer"));

var _rimraf = _interopRequireDefault(require("rimraf"));

var _terminalLink = _interopRequireDefault(require("terminal-link"));

var _internal = require("@redwoodjs/internal");

var _detection = require("@redwoodjs/prerender/detection");

var _lib = require("../lib");

var _colors = _interopRequireDefault(require("../lib/colors"));

var _generatePrismaClient = require("../lib/generatePrismaClient");

var _prerender = require("./prerender");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const command = 'build [side..]';
exports.command = command;
const description = 'Build for production';
exports.description = description;

const builder = yargs => {
  const apiExists = _fs.default.existsSync((0, _lib.getPaths)().api.src);

  const webExists = _fs.default.existsSync((0, _lib.getPaths)().web.src);

  const optionDefault = (apiExists, webExists) => {
    let options = [];

    if (apiExists) {
      options.push('api');
    }

    if (webExists) {
      options.push('web');
    }

    return options;
  };

  yargs.positional('side', {
    choices: ['api', 'web'],
    default: optionDefault(apiExists, webExists),
    description: 'Which side(s) to build',
    type: 'array'
  }).option('stats', {
    default: false,
    description: `Use ${(0, _terminalLink.default)('Webpack Bundle Analyzer', 'https://github.com/webpack-contrib/webpack-bundle-analyzer')}`,
    type: 'boolean'
  }).option('verbose', {
    alias: 'v',
    default: false,
    description: 'Print more',
    type: 'boolean'
  }).option('prerender', {
    default: true,
    description: 'Prerender after building web',
    type: 'boolean'
  }).option('prisma', {
    type: 'boolean',
    alias: 'db',
    default: true,
    description: 'Generate the Prisma client'
  }).option('performance', {
    alias: 'perf',
    type: 'boolean',
    default: false,
    description: 'Measure build performance'
  }).epilogue(`Also see the ${(0, _terminalLink.default)('Redwood CLI Reference', 'https://redwoodjs.com/reference/command-line-interface#build')}`);
};

exports.builder = builder;

const handler = async ({
  side = ['api', 'web'],
  verbose = false,
  performance = false,
  stats = false,
  prisma = true,
  prerender
}) => {
  const rwjsPaths = (0, _lib.getPaths)();

  if (performance) {
    console.log('Measuring Web Build Performance...');

    _execa.default.sync(`yarn cross-env NODE_ENV=production webpack --config ${require.resolve('@redwoodjs/core/config/webpack.perf.js')}`, {
      stdio: 'inherit',
      shell: true
    }); // We do not want to continue building...


    return;
  }

  if (stats) {
    console.log('Building Web Stats...');

    _execa.default.sync(`yarn cross-env NODE_ENV=production webpack --config ${require.resolve('@redwoodjs/core/config/webpack.stats.js')}`, {
      stdio: 'inherit',
      shell: true
    }); // We do not want to continue building...


    return;
  }

  const tasks = [side.includes('api') && prisma && {
    title: 'Generating Prisma Client...',
    task: async () => {
      const {
        cmd,
        args
      } = (0, _generatePrismaClient.generatePrismaCommand)(rwjsPaths.api.dbSchema);
      return (0, _execa.default)(cmd, args, {
        stdio: verbose ? 'inherit' : 'pipe',
        shell: true,
        cwd: rwjsPaths.api.base
      });
    }
  }, side.includes('api') && {
    title: 'Building API...',
    task: () => {
      const {
        errors,
        warnings
      } = (0, _internal.buildApi)();

      if (errors.length) {
        console.error(errors);
      }

      if (warnings.length) {
        console.warn(warnings);
      }
    }
  }, side.includes('web') && {
    // Clean web
    title: 'Cleaning Web...',
    task: () => {
      _rimraf.default.sync(rwjsPaths.web.dist);
    }
  }, side.includes('web') && {
    title: 'Building Web...',
    task: () => {
      return (0, _execa.default)(`yarn cross-env NODE_ENV=production webpack --config ${require.resolve('@redwoodjs/core/config/webpack.production.js')}`, {
        stdio: verbose ? 'inherit' : 'pipe',
        shell: true,
        cwd: rwjsPaths.web.base
      });
    }
  }, side.includes('web') && prerender && {
    title: 'Prerendering Web...',
    task: async () => {
      const prerenderRoutes = (0, _detection.detectPrerenderRoutes)();

      if (prerenderRoutes.length === 0) {
        return `You have not marked any "prerender" in your ${(0, _terminalLink.default)('Routes', 'file://' + rwjsPaths.web.routes)}.`;
      }

      return new _listr.default(await (0, _prerender.getTasks)(), {
        renderer: verbose && _listrVerboseRenderer.default,
        concurrent: true // Re-use prerender tasks, but run them in parallel to speed things up

      });
    }
  }].filter(Boolean);
  const jobs = new _listr.default(tasks, {
    renderer: verbose && _listrVerboseRenderer.default
  });

  try {
    await jobs.run();
  } catch (e) {
    console.log(_colors.default.error(e.message));
    process.exit(1);
  }
};

exports.handler = handler;
