// @redwoodjs/cli/dist/commands/storybook.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.handler = exports.builder = exports.description = exports.aliases = exports.command = void 0;

var _path = _interopRequireDefault(require("path"));

var _execa = _interopRequireDefault(require("execa"));

var _internal = require("@redwoodjs/internal");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const command = 'storybook';
exports.command = command;
const aliases = ['sb'];
exports.aliases = aliases;
const description = 'Launch Storybook: An isolated component development environment';
exports.description = description;

const builder = yargs => {
  yargs.option('open', {
    describe: 'Open storybooks in your browser on start',
    type: 'boolean',
    default: true
  }).option('build', {
    describe: 'Build Storybook',
    type: 'boolean',
    default: false
  }).option('port', {
    describe: 'Which port to run storybooks on',
    type: 'integer',
    default: 7910
  }).option('build-directory', {
    describe: 'Directory in web/ to store static files',
    type: 'string',
    default: 'public/storybook'
  });
};

exports.builder = builder;

const handler = ({
  open,
  port,
  build,
  buildDirectory
}) => {
  const cwd = (0, _internal.getPaths)().web.base;

  const staticAssetsFolder = _path.default.join((0, _internal.getPaths)().web.base, 'public'); // Create the `MockServiceWorker.js` file
  // https://mswjs.io/docs/cli/init


  (0, _execa.default)(`yarn msw init "${staticAssetsFolder}" --no-save`, undefined, {
    stdio: 'inherit',
    shell: true,
    cwd
  });

  const storybookConfig = _path.default.dirname(require.resolve('@redwoodjs/testing/config/storybook/main.js'));

  (0, _execa.default)(`yarn ${build ? 'build' : 'start'}-storybook`, [`--config-dir "${storybookConfig}"`, !build && `--port ${port}`, !build && '--no-version-updates', !build && `--static-dir "${staticAssetsFolder}"`, build && `--output-dir "${_path.default.join((0, _internal.getPaths)().web.base, buildDirectory)}"`, !open && '--ci'].filter(Boolean), {
    stdio: 'inherit',
    shell: true,
    cwd
  });
};

exports.handler = handler;
