// @prisma/engine-core/dist/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.fixBinaryTargets = exports.NApiEngineTypes = exports.printGeneratorConfig = exports.Engine = exports.NAPIEngine = exports.getInternalDatamodelJson = exports.PrismaClientUnknownRequestError = exports.PrismaClientRustPanicError = exports.PrismaClientKnownRequestError = exports.PrismaClientInitializationError = void 0;
var errors_1 = require("./errors");
Object.defineProperty(exports, "PrismaClientInitializationError", { enumerable: true, get: function () { return errors_1.PrismaClientInitializationError; } });
Object.defineProperty(exports, "PrismaClientKnownRequestError", { enumerable: true, get: function () { return errors_1.PrismaClientKnownRequestError; } });
Object.defineProperty(exports, "PrismaClientRustPanicError", { enumerable: true, get: function () { return errors_1.PrismaClientRustPanicError; } });
Object.defineProperty(exports, "PrismaClientUnknownRequestError", { enumerable: true, get: function () { return errors_1.PrismaClientUnknownRequestError; } });
var getInternalDatamodelJson_1 = require("./getInternalDatamodelJson");
Object.defineProperty(exports, "getInternalDatamodelJson", { enumerable: true, get: function () { return getInternalDatamodelJson_1.getInternalDatamodelJson; } });
var NAPIEngine_1 = require("./NAPIEngine");
Object.defineProperty(exports, "NAPIEngine", { enumerable: true, get: function () { return NAPIEngine_1.NAPIEngine; } });
var NodeEngine_1 = require("./NodeEngine");
Object.defineProperty(exports, "Engine", { enumerable: true, get: function () { return NodeEngine_1.NodeEngine; } });
var printGeneratorConfig_1 = require("./printGeneratorConfig");
Object.defineProperty(exports, "printGeneratorConfig", { enumerable: true, get: function () { return printGeneratorConfig_1.printGeneratorConfig; } });
exports.NApiEngineTypes = __importStar(require("./napi-types"));
var util_1 = require("./util");
Object.defineProperty(exports, "fixBinaryTargets", { enumerable: true, get: function () { return util_1.fixBinaryTargets; } });
