// @redwoodjs/prerender/dist/babelPlugins/babel-plugin-redwood-prerender-media-imports.js
"use strict";

var _Object$defineProperty = require("@babel/runtime-corejs3/core-js/object/define-property");

var _interopRequireDefault = require("@babel/runtime-corejs3/helpers/interopRequireDefault").default;

_Object$defineProperty(exports, "__esModule", {
  value: true
});

exports.default = _default;

var _includes = _interopRequireDefault(require("@babel/runtime-corejs3/core-js/instance/includes"));

var _path = require("path");

var _internal = require("@redwoodjs/internal");

const defaultOptions = {
  // This list of extensions matches config for file-loader in
  // packages/core/config/webpack.common.js
  extensions: ['.ico', '.jpg', '.jpeg', '.png', '.gif', '.eot', '.otf', '.webp', '.ttf', '.woff', '.woff2', '.cur', '.ani', '.pdf', '.bmp']
};

function getVariableName(p) {
  var _p$node$specifiers;

  if ((_p$node$specifiers = p.node.specifiers) !== null && _p$node$specifiers !== void 0 && _p$node$specifiers[0] && p.node.specifiers[0].local) {
    return p.node.specifiers[0].local.name;
  }

  return null;
} // Blank single pixel for url-loaded images


const BLANK_IMG_DATA_SRC = 'data:image/gif;base64,R0lGODlhAQABAIAAAP///wAAACH5BAEAAAAALAAAAAABAAEAAAICRAEAOw==';

function _default({
  types: t
}) {
  const manifestPath = (0, _path.join)((0, _internal.getPaths)().web.dist, 'manifest.json');

  const webpackManifest = require(manifestPath);

  return {
    name: 'babel-plugin-redwood-prerender-media-imports',
    visitor: {
      ImportDeclaration(p, state) {
        var _context;

        const ext = (0, _path.extname)(p.node.source.value);
        const options = { ...defaultOptions,
          ...state.opts
        };

        if (ext && (0, _includes.default)(_context = options.extensions).call(_context, ext)) {
          const importConstName = getVariableName(p);
          const webpackManifestKey = `static/media/${(0, _path.basename)(p.node.source.value)}`;
          const processedFilePath = webpackManifest[webpackManifestKey];

          if (importConstName) {
            p.replaceWith(t.variableDeclaration('const', [t.variableDeclarator(t.identifier(importConstName), t.stringLiteral(processedFilePath || BLANK_IMG_DATA_SRC))]));
          }
        }
      }

    }
  };
}
