// @redwoodjs/cli/dist/commands/generate/layout/layout.js
"use strict";

var _Object$defineProperty = require("@babel/runtime-corejs3/core-js/object/define-property");

var _interopRequireDefault = require("@babel/runtime-corejs3/helpers/interopRequireDefault");

_Object$defineProperty(exports, "__esModule", {
  value: true
});

exports.handler = exports.builder = exports.description = exports.command = exports.files = void 0;

var _reduce = _interopRequireDefault(require("@babel/runtime-corejs3/core-js/instance/reduce"));

var _lib = require("../../../lib");

var _generate = require("../../generate");

var _helpers = require("../helpers");

const COMPONENT_SUFFIX = 'Layout';
const REDWOOD_WEB_PATH_NAME = 'layouts';

const files = ({
  name,
  tests = true,
  stories = true,
  ...options
}) => {
  // TODO: Replace with check from https://github.com/redwoodjs/redwood/pull/633
  const isJavascript = options.javascript && !options.typescript;
  const layoutFile = (0, _helpers.templateForComponentFile)({
    name,
    suffix: COMPONENT_SUFFIX,
    webPathSection: REDWOOD_WEB_PATH_NAME,
    extension: isJavascript ? '.js' : '.tsx',
    generator: 'layout',
    templatePath: options.skipLink ? 'layout.tsx.a11yTemplate' : 'layout.tsx.template'
  });
  const testFile = (0, _helpers.templateForComponentFile)({
    name,
    suffix: COMPONENT_SUFFIX,
    extension: `.test.${isJavascript ? 'js' : 'tsx'}`,
    webPathSection: REDWOOD_WEB_PATH_NAME,
    generator: 'layout',
    templatePath: 'test.tsx.template'
  });
  const storyFile = (0, _helpers.templateForComponentFile)({
    name,
    suffix: COMPONENT_SUFFIX,
    extension: `.stories.${isJavascript ? 'js' : 'tsx'}`,
    webPathSection: REDWOOD_WEB_PATH_NAME,
    generator: 'layout',
    templatePath: 'stories.tsx.template'
  });
  const files = [layoutFile];

  if (stories) {
    files.push(storyFile);
  }

  if (tests) {
    files.push(testFile);
  } // Returns
  // {
  //    "path/to/fileA": "<<<template>>>",
  //    "path/to/fileB": "<<<template>>>",
  // }


  return (0, _reduce.default)(files).call(files, (acc, [outputPath, content]) => {
    const template = isJavascript ? (0, _lib.transformTSToJS)(outputPath, content) : content;
    return {
      [outputPath]: template,
      ...acc
    };
  }, {});
};

exports.files = files;
const builderObj = {
  skipLink: {
    default: false,
    description: 'Generate with skip link',
    type: 'boolean'
  },
  ..._generate.yargsDefaults
};
const {
  command,
  description,
  builder,
  handler
} = (0, _helpers.createYargsForComponentGeneration)({
  componentName: 'layout',
  filesFn: files,
  builderObj
});
exports.handler = handler;
exports.builder = builder;
exports.description = description;
exports.command = command;
