// @redwoodjs/api/dist/makeMergedSchema/rootSchema.js
"use strict";

var _Object$defineProperty = require("@babel/runtime-corejs3/core-js/object/define-property");

var _interopRequireDefault = require("@babel/runtime-corejs3/helpers/interopRequireDefault").default;

_Object$defineProperty(exports, "__esModule", {
  value: true
});

exports.resolvers = exports.schema = void 0;

var _graphqlScalars = require("graphql-scalars");

var _graphqlTag = _interopRequireDefault(require("graphql-tag"));

var _package = _interopRequireDefault(require("../../package.json"));

/* eslint-disable @typescript-eslint/ban-ts-comment */
// eslint-disable-next-line @typescript-eslint/ban-ts-comment
// @ts-ignore - not inside the <rootDir>
const {
  prismaVersion
} = require('@prisma/client');
/**
 * This adds scalar types for dealing with Date, Time, DateTime, and JSON.
 * This also adds a root Query type which is needed to start the GraphQL server on a fresh install.
 */


const schema = (0, _graphqlTag.default)`
  scalar Date
  scalar Time
  scalar DateTime
  scalar JSON
  scalar JSONObject

  type Redwood {
    version: String
    currentUser: JSON
    prismaVersion: String
  }

  type Query {
    redwood: Redwood
  }
`;
exports.schema = schema;
const resolvers = {
  Date: _graphqlScalars.DateResolver,
  Time: _graphqlScalars.TimeResolver,
  DateTime: _graphqlScalars.DateTimeResolver,
  JSON: _graphqlScalars.JSONResolver,
  JSONObject: _graphqlScalars.JSONObjectResolver,
  Query: {
    redwood: () => ({
      version: _package.default.version,
      prismaVersion: () => prismaVersion.client,
      currentUser: (_args, context) => {
        return context === null || context === void 0 ? void 0 : context.currentUser;
      }
    })
  }
};
exports.resolvers = resolvers;
