// @redwoodjs/structure/dist/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getProject = getProject;
exports.printDiagnostics = printDiagnostics;
Object.defineProperty(exports, "DefaultHost", {
  enumerable: true,
  get: function () {
    return _hosts.DefaultHost;
  }
});
Object.defineProperty(exports, "Host", {
  enumerable: true,
  get: function () {
    return _hosts.Host;
  }
});
Object.defineProperty(exports, "RWProject", {
  enumerable: true,
  get: function () {
    return _model.RWProject;
  }
});
Object.defineProperty(exports, "DiagnosticSeverity", {
  enumerable: true,
  get: function () {
    return _vscodeLanguageserverTypes2.DiagnosticSeverity;
  }
});
Object.defineProperty(exports, "URL_file", {
  enumerable: true,
  get: function () {
    return _URL.URL_file;
  }
});

var _hosts = require("./hosts");

var _model = require("./model");

var _vscodeLanguageserverTypes = require("./x/vscode-languageserver-types");

var _vscodeLanguageserverTypes2 = require("vscode-languageserver-types");

var _URL = require("./x/URL");

function getProject(projectRoot, host = new _hosts.DefaultHost()) {
  return new _model.RWProject({
    projectRoot,
    host
  });
}

async function printDiagnostics(projectRoot, opts) {
  const project = getProject(projectRoot);
  const formatOpts = {
    cwd: projectRoot,
    ...opts
  };

  try {
    let exceptions = 0;

    for (const d of await project.collectDiagnostics()) {
      const str = (0, _vscodeLanguageserverTypes.ExtendedDiagnostic_format)(d, formatOpts);
      console.log(str); // counts number of warnings (2) and errors (1) encountered

      if (d.diagnostic.severity === 2 || d.diagnostic.severity === 1) {
        exceptions++;
      }
    }

    if (exceptions === 0) {
      console.log('Success: no errors or warnings were detected');
    }
  } catch (e) {
    console.log('runtime error: ' + e.message);
  }
}
