// @redwoodjs/prerender/dist/internal.js
"use strict";

var _Object$defineProperty = require("@babel/runtime-corejs3/core-js/object/define-property");

var _interopRequireDefault = require("@babel/runtime-corejs3/helpers/interopRequireDefault");

_Object$defineProperty(exports, "__esModule", {
  value: true
});

exports.writeToDist = exports.registerShims = exports.getRootHtmlPath = void 0;

var _fs = _interopRequireDefault(require("fs"));

var _path = _interopRequireDefault(require("path"));

var _nodeFetch = _interopRequireDefault(require("node-fetch"));

var _internal = require("@redwoodjs/internal");

const INDEX_FILE = _path.default.join((0, _internal.getPaths)().web.dist, 'index.html');

const DEFAULT_INDEX = _path.default.join((0, _internal.getPaths)().web.dist, 'defaultIndex.html');

const getRootHtmlPath = () => {
  if (_fs.default.existsSync(DEFAULT_INDEX)) {
    return DEFAULT_INDEX;
  } else {
    return INDEX_FILE;
  }
};

exports.getRootHtmlPath = getRootHtmlPath;

const registerShims = () => {
  global.__REDWOOD__API_PROXY_PATH = (0, _internal.getConfig)().web.apiProxyPath;

  global.__REDWOOD__USE_AUTH = () => ({
    loading: true,
    // this should play nicely if the app waits for auth stuff to comeback first before render
    isAuthenticated: false
  }); // we only need a parital AuthContextInterface for prerender


  global.__REDWOOD__PRERENDERING = true; // Let routes auto loader plugin know

  process.env.__REDWOOD__PRERENDERING = '1'; // Shim fetch in the node.js context
  // This is to avoid using cross-fetch when configuring apollo-client
  // which would cause the client bundle size to increase

  if (!global.fetch) {
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-ignore-next-line
    global.fetch = _nodeFetch.default;
  }
};

exports.registerShims = registerShims;

const writeToDist = (outputHtmlPath, renderOutput) => {
  const dirName = _path.default.dirname(outputHtmlPath);

  const exist = _fs.default.existsSync(dirName);

  if (!exist) {
    _fs.default.mkdirSync(dirName, {
      recursive: true
    });
  }

  _fs.default.writeFileSync(outputHtmlPath, renderOutput);
};

exports.writeToDist = writeToDist;
